mokapot/confidence.py

```diff
@@ -359,10 +405,16 @@ def create_output_writer(path: Path, level: str, initialize: bool):
             writer.initialize()
         return writer
 
-    for dataset, score, prefix in zip(datasets, scores_list, prefixes):
+    if prefixes is None:
+        prefixes = [None] * len(datasets)
+
+    out = []
+
+    for dataset, score, prefix in strictzip(datasets, scores_list, prefixes):
         # todo: nice to have: move this column renaming stuff into the
         #   column defs module, and further, have standardized columns
         #   directly from the pin reader (applying the renaming itself)
+
         level_column_names = [
             "PSMId",
```

Comment:
Question (@ezander and @gessulat) Is it a dealbreaker for you to have this column hard-coded? I believe that the current behavior in the published version of mokapot is to preserve all the "spectrum columns" (which usually will be psm id + file name + rank ... all columns that uniquely identify a PSM). Let me know if having a single "PSMId" column is a requirement.

Reply:
Hi @jspaezp 
I looked again at the current state of column handling in mokapot, which is a little bit confusing. This may not answer your question directly, but may give a basis for moving forward.

Input columns are determined by parsers/pin.py
* All are case-insensitive
* Required
    * specid (becomes later PSMId, not identical to the "spectrum")
    * peptide
    * proteins
    * label (target column)
    * scannr
* Optional
    * rollup columns: modifiedpeptides, precursor, peptidegroup
    * filename
    * calcmass
    * expmass
    * ret_time
    * charge
* Spectra is [filename, scannr, ret_time, expmass]. Since all except scannr are optional, can
  span between 1 and 4 columns.
  Is used for two things (IIRC): in model training for competition (only those with same "spectrum" compete) and later
  in the deduplication process in confidence writing.
* From this an OnDiskPsmDataset is created and the names of the inferred columns are set
* IMHO the dataset should rather get a reader, and the plain CSVFileReader should be wrapped in a ColumnMappedReader so
  that in the code, only hard-coded names (or names defined as constants) can be used (except maybe for the feature
  columns, but there could also be some mapping scheme e.g. from "foo" -> "feature_001", "bar" -> "feature_002", ...)
* Also: any conversion necessary for the input file should be done here (e.g. convert label from int to bool)

The `level_columns` (yeah, the naming...) are the columns of the intermediate files, that get written for each rollup
level before confidence assignment is made.
* During writing of the "level files" deduplication is done on the columns as given by the "spectra columns" i.e. some
  subset of [filename, scannr, ret_time, expmass]
* Columns are renamed from the `level_input_column_names` to the `level_column_names`. Most importantly the "specId"
  column is renamed to "PSMId".
  (It's not quite clear why the "target_column" is not renamed, probably the name taken from the input file is usually
  correct by some coincidence). This whole renaming business (as mentions should IMO move to the input file reader as
  mentioned earlier)

`output_column_names` are the names of the columns after confidence estimation on the "(rollup) level files".
* The names are hard-coded except for the `extra_output_columns`, which (as of now) only contain the optional rollup
  levels. One more point for defining those as string constants somewhere.
* Additionally there are the `score`, `q-value`, and `posterior_error_prob` columns. The `-` is problematic when writing
  to SQL (should be q_value, but that clashed with other code momentarily)
* The names of the columns could be inferred from reader columns when writing the output files. However, due to the
  possibility of multiple input files going into one output file (and the inherited code structure) it was necessary
  to write the column header before they could be inferred from a reader (but this could be changed, of course).

Note: it is still somewhat messy, but can be cleared up. The special treatment of proteins (which, according to some,
shouldn't be mokapot's business anyway) makes that somewhat harder.

Reply:
> Columns are renamed from the level_input_column_names to the level_column_names. Most importantly the "specId"
column is renamed to "PSMId".

This assumes there is a single column that uniquely identifies the spectrum. Since we are not writing say ... the file name, there is the clear assumption that the information would be redundant with the `PSMId`.

I believe that the point I want to make is... SINCE we are requesting and supporting multiple columns denoting uniqueness of the spectrum why are we not writing them? IMHO If the user wants a single column called `PSMId`, they should pass it as an input. (input columns should be the same as output columns)

> IMHO the dataset should rather get a reader, and the plain CSVFileReader should be wrapped in a ColumnMappedReader so
that in the code, only hard-coded names (or names defined as constants) can be used (except maybe for the feature
columns, but there could also be some mapping scheme e.g. from "foo" -> "feature_001", "bar" -> "feature_002", ...)

I don't think it matters, the interface should be defined in the ABC, not in the `OnDiskPSMDataset`

> Additionally there are the score, q-value, and posterior_error_prob columns. The - is problematic when writing

Cool, I will make sure columns are sql-friendly.

> The names of the columns could be inferred from reader columns when writing the output files. However, due to the
possibility of multiple input files going into one output file (and the inherited code structure) it was necessary
to write the column header before they could be inferred from a reader (but this could be changed, of course).

But we check when multiple datasets are passed that all of them have the same columns ... (I dont think we should support multiple input files where each can have  different column names)

## My proposal:

```
Target/Label_column: [str]
	Column that marks whether an element is a target/decoy.
	Bool? (I think it has to be a bool RN)

Spectrum_columns: list[str]
	The series of columns that denote which PSMs are competing with each other. For example: 
        1. We can have both a target and a decoy coming from the same scan number + file.
        2. We can have multiple peptides suggested as targets for each scan.
	We should also warn if a float column is passed here, since testing equality in floats is not reliable.

Peptide_column: [str]
	A column that denotes a peptide, this column + spectrum_columns should identify uniquely a PSM (thus have no duplicates)

Extra rollup columns: list[str]
	Other columns that can be used to compete the spectra for summary.
	Each of those columns will be used to summarize the output as at a different level.
	For example: modifiedpeptides, precursor, peptidegroup
	
Feature_columns: list[str]
	List of all the columns that can be used as possible scores to build a model.

```

1. Trim the PSMDataset API to contain only those groups (that includes the LinearPSMDataset and the OnDiskPsmDataset).
2. Request other columns as needed for other outputs (move to ‘to_flashlfq’ or so …)
3. Store the best scores internally in the Dataset (instead of being passed by `brew` and then used again in the confidence estimation)
4. Write columns as-is (maybe replace spaces + punctuation/hyphenation to '_')